File: 백준/Silver/15657. N과 M （8）/N과 M （8）.cc
#include <bits/stdc++.h>

using namespace std;


int n, m;
vector<int> arr;
vector<int> v;

void bt(int cnt, int idx)
{
	if (cnt == m)
	{
		for (int i = 0; i < v.size(); i++)
		{
			cout << v[i] << " ";
		}
		cout << "\n";
		return;
	}

	for (int i = idx; i < n; i++)
	{
		v.push_back(arr[i]);
		bt(cnt + 1, i);
		v.pop_back();
	}
}

int main()
{
	cin >> n >> m;

	for (int i = 0; i < n; i++)
	{
		int temp;
		cin >> temp;
		arr.push_back(temp);
	}

	sort(arr.begin(), arr.end());
	
	bt(0, 0);


	return 0;
}

File: 백준/Silver/15657. N과 M （8）/N과 M （8）.java
import java.io.*;
import java.util.*;

public class Main {
	
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static StringBuilder sb = new StringBuilder();
	static int n, m;
	static int[] arr, result;
	public static void main(String[] args) throws IOException {
		
		st = new StringTokenizer(br.readLine());
		
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		
		arr = new int[n];
		result = new int[n];
		
		st = new StringTokenizer(br.readLine());
		
		for (int i = 0 ; i < n ; i ++) {
			arr[i] = Integer.parseInt(st.nextToken()); 
		}
		
		Arrays.sort(arr);
		
		combi(0, 0);
		
		System.out.print(sb);
		
	}
	
	static void combi(int cnt, int idx) {
		if (cnt == m) {
			
			for (int i = 0 ; i < m; i++) {
				sb.append(result[i] + " ");
			}
			sb.append("\n");
			
			return;
		}
		
		for (int i = idx ; i < n ; i ++ ) {
			result[cnt] = arr[i]; 
			combi(cnt+1, i);
		}
	}
	
}
